lib/pmdtester/builders/rule_set_builder.rb

```diff
@@ -0,0 +1,107 @@
+# frozen_string_literal: true
+
+require 'nokogiri'
+require 'set'
+require_relative '../cmd'
+module PmdTester
+  # This class is responsible for generation dynamic configuration
+  # according to the difference between base and patch branch of Pmd.
+  class RuleSetBuilder
+    ALL_RULE_SETS = Set['bestpractices', 'codestyle', 'design', 'documentation', 'errorprone',
+                        'migrating', 'multithreading', 'performance', 'regex', 'security'].freeze
+    PATH_TO_PMD_JAVA_BASED_RULES =
+      'pmd-java/src/main/java/net/sourceforge/pmd/lang/java/rule/'
+    PATH_TO_PMD_XPATH_BASED_RULES = 'pmd-java/src/main/resources/category/java/'
+    PATH_TO_ALL_JAVA_RULES = 'config/all-java.xml'
+    PATH_TO_DYNAMIC_CONFIG = 'target/dynamic-config.xml'
+    NO_JAVA_RULES_CHANGED_MESSAGE = 'No java rules have been changed!'
+
+    def initialize(options)
+      @options = options
+    end
+
+    def build
+      filenames = diff_filenames
+      rule_sets = get_rule_sets(filenames)
+      build_config_file(rule_sets)
+    end
+
+    def filter_set(rule_sets)
+      if rule_sets.size.eql?(ALL_RULE_SETS.size)
+        # if `rule_sets` contains all rule sets, than no need to filter the baseline
+        @options.filter_set = nil
+      else
+        @options = rule_sets
+      end
+    end
+
+    def diff_filenames
+      filenames = nil
+      Dir.chdir(@options.local_git_repo) do
+        diff_cmd = "git diff --name-only #{@options.base_branch}..#{@options.patch_branch}"
+        grep_cmd = "grep -E '^pmd-(core|java)/'" # OPTIMIZE: extend other languages
+        filenames = Cmd.execute("#{diff_cmd} | #{grep_cmd}")
+      end
+      filenames.split("\n")
+    end
+
+    def get_rule_sets(filenames)
+      rule_sets = Set[]
+      filenames.each do |filename|
+        if filename.start_with?(PATH_TO_PMD_JAVA_BASED_RULES, PATH_TO_PMD_XPATH_BASED_RULES)
+          add_rule_set(rule_sets, filename)
+        else
+          rule_sets = ALL_RULE_SETS
+          break
+        end
+      end
+      rule_sets
+    end
+
+    def add_rule_set(rule_sets, filename)
+      ALL_RULE_SETS.each do |rule_set|
+        unless filename.index(rule_set).nil?
+          rule_sets.add(rule_set)
+          break
+        end
+      end
+    end
+
+    def build_config_file(rule_sets)
+      if rule_sets.empty?
+        puts NO_JAVA_RULES_CHANGED_MESSAGE
+        exit 0
+      end
+
+      doc = Nokogiri::XML(File.read(PATH_TO_ALL_JAVA_RULES))
+      doc.search('rule').each do |rule|
+        rule.remove unless match_ref?(rule, rule_sets)
```

Comment:
Since we are dealing now on the level of complete rulesets, you could easily generate out a new ruleset referencing only the changes categories:

Pseudo code:
```
rule_sets.each do |rule_set|
    puts "<rule ref=\"category/java/#{rule_set}.xml\">"
end
```

We can reference the complete rulesets/categories. You'll need of course the surrounding `<ruleset>` tag and the XML prolog. And maybe use a library to create the xml...

That way, we avoid looping through all the rules and repeatedly through the selected ruleset...

Reply:
Ok, didn't know that PATH_TO_ALL_JAVA_RULES already contains the references to the complete rulesets, so these two nested loops are not too big actually. Still I'd try to avoid too many loopings...